src/Cardano/Wallet/CoinSelection/Fee.hs

```diff
@@ -0,0 +1,351 @@
+            -- change outputs, we make them smaller and may reduce the size of the
+            -- transaction, and the fee. Thus, we end up paying slightly more than
+            -- the upper bound. We could do some binary search and try to
+            -- re-distribute excess across changes until fee becomes bigger.
+            if (getFee remainingFee) == Coin 0 then
+                pure $ CoinSelection inps outs chgs'
+            else do
+            -- 3.2/
+            -- Otherwise, we need an extra entries from the available utxo to
+            -- cover what's left. Note that this entry may increase our change
+            -- because we may not consume it entirely. So we will just split
+            -- the extra change across all changes possibly increasing the
+            -- number of change outputs (if there was none, or if increasing
+            -- a change value causes an overflow).
+            --
+            -- Because selecting a new input increases the fee, we need to
+            -- re-run the algorithm with this new elements and using the initial
+            -- change plus the extra change brought up by this entry and see if
+            -- we can now correctly cover fee.
+                remFee <- lift $ runMaybeT $ coverRemainingFee remainingFee utxo
+                case remFee of
+                    Nothing ->
+                        throwE CannotCoverFee
+                    Just inps' -> do
+                        let extraChange = splitChange (currentBalance inps') chgs
+                        pure $ CoinSelection (inps <> inps') outs extraChange
+
+feeUpperBound
+    :: FeeOptions
+    -> CoinSelection
+    -> Fee
+feeUpperBound opt (CoinSelection inps outs chgs) =
+    (estimate opt) (L.length inps) (map coin outs ++ chgs)
+
+-- | Proportionally divide the fee over each output.
+-- Pre-condition 1: The given outputs list shouldn't be empty
+-- Pre-condition 2: None of the outputs should be null
+divvyFee
+    :: Fee
+    -> [Coin]
+    -> [(Fee, Coin)]
+divvyFee _ [] =
+    error "divvyFee: empty list"
+divvyFee _ outs | (Coin 0) `elem` outs =
+    error "divvyFee: some outputs are null"
+divvyFee fee outs =
+    map (\a -> (feeForOut a, a)) outs
+  where
+    totalOut :: Coin
+    totalOut = (Coin . sum . map getCoin) outs
+
+    -- The ratio will be between 0 and 1 so cannot overflow
+    feeForOut :: Coin -> Fee
+    feeForOut a =
+        let res = valueAdjust (valueRatio a totalOut) (getFee fee)
+        in if isValidCoin res then
+            Fee res
+        else
+            error "feeForOut : fee exceeded maximum valid value for Coin"
+
+
+    valueRatio :: Coin -> Coin -> Double
+    valueRatio c1 c2 = (coinToDouble c1) / (coinToDouble c2)
+
+    valueAdjust
+        :: Double
+        -> Coin
+        -> Coin
+    valueAdjust d c =
+        Coin $ ceiling (d * (coinToDouble c))
+
+
+coinToDouble :: Coin -> Double
+coinToDouble = fromIntegral . getCoin
```

Comment:
We can probably simplify those above and not actually require `Double` here. We do know the underlying type for coins (`Word64`), so let's use that and arithmetic on integer (e.g. `div` or `quot` instead of `/`)